### DISCUSSION TITLE
Bug with the refactored `AssetSourceGenerator` in 5.96.0

### DISCUSSION THREAD

**Discussion by titouanmathis** [webpack/webpack#18949] [Q&A] [answered]
# Bug report

**What is the current behavior?**

When using the html-webpack-plugin with a loader to transform its templates, I get the following error : 

```
Error: The loader "/path/to/index.html" didn't return html.
```

**If the current behavior is a bug, please provide the steps to reproduce.**

See https://github.com/titouanmathis/repro-webpack-html-plugin-asset-source-regression for a reproduction.

This seems to be a Webpack regression, so I have not yet opened an issue in the html-webpack-plugin repository.

The reproduction repository contains a patch, but I am not sure if it is the best way to fix this regression.

I was able to identify a7e5e6a02b6a6ab761ad396fdf5fe780ca80c86f as the first bad commit with git bisect. 

**What is the expected behavior?**

The loader should be applied and return the transformed content instead of an empty string.

**Other relevant information:**
webpack version: 5.96.0  
Node.js version: 22
Operating System: macOS
Additional tools: html-webpack-plugin

**Comment by alexander-akait** [comment, marked as answer]
https://github.com/webpack/webpack/issues/18928

**Comment by alexander-akait** [comment]
So this is the expected result, use `type: "javascript/auto"` instead `type: "asset/source"`

**Comment by titouanmathis** [comment]
Thanks for the quick response, I missed the existing issue when searching for my error. I will update the loader on my end. 

Sorry for the noise.

**Reply by alexander-akait** [reply]
I forget to write code for loader:
```
export default function loader(source) {
  return 'export default ' + JSON.stringify(source);
}
```

**Reply by alexander-akait** [reply]
```
 {
    test: /\.html/,
    type: 'javascript/auto',
    loader: './loaders/html.js',
  },
```
